workspaces/translations/plugins/translations-backend/src/service/TranslationService.ts

```diff
@@ -0,0 +1,315 @@
+/*
+ * Copyright Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { Config } from '@backstage/config';
+import fs from 'fs';
+import path from 'path';
+
+import {
+  deepMergeTranslations,
+  filterLocales,
+  isValidJSONTranslation,
+} from '../utils';
+
+export interface TranslationFileStats {
+  filesProcessed: number;
+  filesNotFound: number;
+  filesInvalid: number;
+}
+
+export interface TranslationResult {
+  translations: Record<string, any>;
+  stats: {
+    totalFilesProcessed: number;
+    totalFilesNotFound: number;
+    totalFilesInvalid: number;
+  };
+}
+
+/**
+ * Service for managing translation files and processing
+ */
+export class TranslationService {
+  private cachedTranslations: Record<string, any> | null = null;
+  private readonly overridesFiles: string[];
+  private readonly configuredLocales: string[];
+
+  constructor(
+    private readonly config: Config,
+    private readonly logger: LoggerService,
+  ) {
+    this.overridesFiles =
+      this.config.getOptionalStringArray('i18n.overrides') ?? [];
+    this.configuredLocales = this.config.getOptionalStringArray(
+      'i18n.locales',
+    ) ?? ['en'];
+  }
+
+  /**
+   * Finds the topmost folder with the given name by searching up the directory tree
+   * Similar to the approach used in marketplace plugin
+   */
+  private findTopmostFolder(
+    folderName: string,
+    startPath = process.cwd(),
+  ): string | null {
+    let currentPath = path.resolve(startPath);
+    let topmostFoundPath: string | null = null;
+
+    while (currentPath !== path.parse(currentPath).root) {
+      const targetFolderPath = path.join(currentPath, folderName);
+      if (
+        fs.existsSync(targetFolderPath) &&
+        fs.statSync(targetFolderPath).isDirectory()
+      ) {
+        topmostFoundPath = targetFolderPath;
+      }
+      currentPath = path.dirname(currentPath);
+    }
+    const targetFolderPath = path.join(currentPath, folderName);
+
+    if (
+      fs.existsSync(targetFolderPath) &&
+      fs.statSync(targetFolderPath).isDirectory()
+    ) {
+      topmostFoundPath = targetFolderPath;
+    }
+
+    return topmostFoundPath;
+  }
```

Comment:
I'm not a fan of this "magic" but its fine for now. I like to discuss it next week.

I would prefer that we configure these paths in the app-config.

Reply:
Sure, we can improve this next week in a follow up PR.